lib/install/deps.js

```diff
@@ -65,6 +65,8 @@ function doesChildVersionMatch (child, requested, requestor) {
   // we always consider deps provided by a shrinkwrap as "correct" or else
   // we'll subvert them if they're intentionally "invalid"
   if (child.parent === requestor && child.fromShrinkwrap) return true
+  // symlinked packages without a version specifier should also be assumed to be "correct"
+  if (child.parent === requestor && child.isLink && !registryTypes[requested.type]) return true
```

Comment:
Point of discussion:  If my dependency *does* have a version specifier in `package.json`, we'll still traverse it if the versions do not match.

Should the presence of a symlinked package take precedence over whatever is defined in package.json? 

---

I haven't been able to reproduce the behavior described in #10343 with a versioned dependency, but if a conflict exists, we print a warning, as well as a _ton_ of spurious warning messages about redundant dependencies existing inside of the symlinked module.  

For instance, if I `npm link` `lodash@4.0.0` into a package that asks for `lodash@2.0.0` in its package.json, I'll get several hundred messages along the lines of:

```
npm WARN skippingAction Module is inside a symlinked module: not running remove fs-extra@0.30.0 node_modules/lodash/node_modules/fs-extra
npm WARN skippingAction Module is inside a symlinked module: not running remove fileset@0.2.1 node_modules/lodash/node_modules/fileset
npm WARN skippingAction Module is inside a symlinked module: not running remove xmlbuilder@3.1.0 node_modules/lodash/node_modules/xmlbuilder
...
npm WARN update-linked node_modules/lodash needs updating to 2.0.0 from 4.13.2-pre but we can't, as it's a symlink
```

Removing the `&& !registryTypes[requested.type]` from the conditional above eliminates these messages, which is _probably_ what we want here.  Right?